apps/nextra/pages/en/build/smart-contracts/book/enums.mdx

```diff
@@ -0,0 +1,214 @@
+# Enums
+
+_Since language version 2.0_
+
+Enum types are similar to struct types but support defining multiple *variants* of the data layout. Each variant has its distinct set of fields. In expressions, tools for testing, matching, and deconstructing enum variants are supported. 
+
+
+## Declaration of Enum Types 
+
+An enum type declaration lists the number of different variants, as seen in the example below:
+
+```move
+enum Shape {
+    Circle{radius: u64},
+    Rectangle{width: u64, height: u64}
+}
+```
+
+There can be zero or more fields for an enum variant. If no arguments are given, the braces can also be omitted, declaring simple values:
+
+```move
+enum Color {
+  Red, Blue, Green
+}
+```
+
+Similar to struct types, enum types can have abilities. For example, for the `Color` enum type it is more appropriate to make it copyable, droppable, and storable, similar like primitive number types:
+
+```move
+enum Color has copy, drop, store, key { ... }
+```
+
+Enum types can also have the key ability and appear as roots of data in global storage. A common usage of enums in this context is versioning of data:
+
+```move
+enum VersionedData has key {
+  V1{name: String}
+  V2{name: String, age: u64}
+}
+```
+
+Similar to structs, enum types can be generic. For example, the type below represents a generic result type, where the variant constructors use positional instead of named arguments (see also [positional structs](structs-and-resources.mdx#positional-structs)).
+
+```move
+enum Result<T> has copy, drop, store {
+  Err(u64),
+  Ok(T)
+}
+```
+
+## Constructing Enum Values 
+
+An enum value is constructed similarly to a struct value:
+
+```move
+let s: String;
+let data = VersionedData::V1{name: s};
+```
+
+If the enum variant has no fields, the braces can also be omitted:
+
+```move
+let color = Color::Blue;
+```
+
+## Name Resolution for Enum Variants 
+
+An enum variant names need to be qualified by the enum type name, as in `VersionedData::V1`. 
+
+> Note:" Aliasing via the `use` clause is currently not supported for enum variants, and will be added in later language versions
+
+In selected contexts, the Move compiler infers the enum type from the context, and the qualification by the type name can be omitted:
+
+```move
+fun f(data: VersionedData) {
+  match (data) { V1{..} => .., ..} // simple variant name OK
+}
+```
+
+## Matching Enum Values 
+
+The value of an enum value can be inspected using a match expression. For example:
+
+```move
+fun area(self: &Rectangle): u64 {
+    match (self) {
+        Circle{radius}           => mul_with_pi(*radius * *radius),
+        Rectangle{width, height) => *width * *height
+    }
+}
+```
+
+Notice that above the value matched is an immutable reference to an enum value. We can also  consume values, or match over a mutable reference for interior updates:
+
+```move
+fun scale_radius(self: &mut Rectanle, factor:  u64) {
+    match (self) {
+        Circle{radius} => *radius = *radius * factor,
```

Comment:
Can we have an example (or reuse this example) to show that we can also do `Circle {radius: r}` in the match arm? This syntax may not be obvious to non-Rust folks (and it might not be obvious how else to avoid shadowing variables).

Reply:
Done